src/peft/tuners/lora/variants.py

```diff
@@ -315,3 +317,162 @@ class DoraConv3dVariant(_DoraConvNdVariant):
+            module.sinelora_scaling = math.sqrt(module.in_features)
+
+    @staticmethod
+    def unmerge(module: Linear, active_adapter: str, orig_weight: torch.Tensor) -> torch.Tensor:
+        orig_dtype = orig_weight.dtype
+        lora_A = module.lora_A[active_adapter]
+        lora_B = module.lora_B[active_adapter]
+        lora_scaling = module.scaling[active_adapter]
+        delta_weight = (
+            torch.sin(module.sinelora_frequency * lora_A.weight.T @ lora_B.weight.T).T
+            / module.sinelora_scaling
+            * lora_scaling
+        )
+
+        delta_weight = delta_weight.to(orig_dtype)
+        unmerged_weight = orig_weight - delta_weight
+        return unmerged_weight
+    
+    
+    @staticmethod
+    def merge_safe(module: Linear, active_adapter: str, orig_weight: torch.Tensor) -> torch.Tensor:
+        orig_dtype = orig_weight.dtype
+        lora_A = module.lora_A[active_adapter]
+        lora_B = module.lora_B[active_adapter]
+        lora_scaling = module.scaling[active_adapter]
+        delta_weight = (
+            torch.sin(module.sinelora_frequency * (lora_A.weight.T @ lora_B.weight.T)).T
+            / module.sinelora_scaling
+            * lora_scaling
+        )
+        merged_weight = orig_weight + delta_weight
+        if not torch.isfinite(merged_weight).all():
+            raise ValueError(f"NaNs detected in merged weights for adapter {active_adapter}")
+        module._cache_store(f"{active_adapter}-delta_weight", delta_weight)
+        merged_weight = merged_weight.to(orig_dtype)
+        return merged_weight
+    @staticmethod
+    def merge_unsafe(module: Linear, active_adapter: str, orig_weight: torch.Tensor) -> None:
+        lora_A = module.lora_A[active_adapter]
+        lora_B = module.lora_B[active_adapter]
+        lora_scaling = module.scaling[active_adapter]
+        delta_weight = (
+            torch.sin(module.sinelora_frequency * (lora_A.weight.T @ lora_B.weight.T)).T
+            / module.sinelora_scaling
+            * lora_scaling
+        )
+        module._cache_store(f"{active_adapter}-delta_weight", delta_weight)
+        orig_weight.data += delta_weight
+    @staticmethod
+    def forward(module: Linear, active_adapter: str, x: torch.Tensor, result: torch.Tensor) -> torch.Tensor:
+        lora_A = module.lora_A[active_adapter]
+        lora_B = module.lora_B[active_adapter]
+        lora_scaling = module.scaling[active_adapter]
+        sine_output = (
+            x
+            @ torch.sin(module.sinelora_frequency * lora_A.weight.T @ lora_B.weight.T)
+            / module.sinelora_scaling
+            * lora_scaling
+        )
+        result = result + sine_output
+        return result
+
+class SineLoraEmbeddingVariant(LoraVariant):
+    @staticmethod
+    def init(module: Embedding, adapter_name: str, **kwargs) -> None:
+        module.sinelora_frequency = kwargs["sinelora_frequency"]
+
+        sinelora_scaling = kwargs["sinelora_scaling"]
```

Comment:
Should probably be `self.sinelora_scaling = kwargs...`.

Please add a test case for this scenario, e.g.:

```
    ("Embedding + transformers Conv1D LoRA + SineLoRA 2", "EmbConv1D", LoraConfig, {"target_modules": ["emb"], "use_sinelora": True, "sinelora_scaling": 100}),
```